src/rt/cpp/when.h

```diff
@@ -0,0 +1,125 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+
+#include "cown.h"
+
+#include <functional>
+#include <tuple>
+#include <utility>
+#include <verona.h>
+
+/**
+ * Class for staging the when creation.
+ *
+ * Do not call directly use `when`
+ *
+ * This provides an operator << to apply the closure.  This allows the
+ * argument order to be more sensible, as variadic arguments have to be last.
+ *
+ *   when (cown1, ..., cownn) << closure;
+ *
+ * Allows the variadic number of cowns to occur before the closure.
+ */
+template<typename... Args>
+class When
+{
+  template<typename... Args2>
+  friend When<Args2...> when(Args2... args);

```

Comment:
Should this be friends with any `when` instantiation or only the one with `Args...`?